### PR TITLE
Delegation abort

### PR DESCRIPTION
**Pull request by tobyhede**
## Summary by CodeRabbit

* **New Features**
  * Added abort command to cancel delegation tokens (supports --force) plus claim and delegate commands; status view now shows active delegations and token hints.

* **Public API / Schemas**
  * Added abort response and delegation status schemas and new user-facing error codes for already-claimed/resolved delegations.

* **Tests**
  * Broad unit and integration coverage for delegation flows, abort/claim edge cases, rendering, and token formatting.

### CODE REVIEW

File: packages/claude-code-plugin/__tests__/config.test.ts [hunk 1]
```diff
@@ -300,3 +300,192 @@ describe('Gate Config Validation', () => {
...
+      gates: 'not an object',
+    };
+
+    await fs.writeFile(path.join(testDir, 'rundown-plugin.json'), JSON.stringify(configObj));
+
+    // Object.entries on a string iterates character indices without throwing
+    // This documents the actual behavior - no runtime error but malformed config
+    const config = await loadConfig(testDir);
+    expect(config).not.toBeNull();
+  });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/claude-code-plugin/__tests__/config.test.ts]: Addressed - test now asserts rejects.toThrow() for non-object gates config.

File: packages/claude-code-plugin/__tests__/config.test.ts [hunk 2]
```diff
@@ -300,3 +300,192 @@ describe('Gate Config Validation', () => {
...
+      },
+    };
+
+    await fs.writeFile(path.join(testDir, 'rundown-plugin.json'), JSON.stringify(configObj));
+
+    // Note: Circular detection happens at runtime in dispatcher, not at config load time
+    // This test documents that config loading doesn't prevent circular references
+    const config = await loadConfig(testDir);
+    expect(config).not.toBeNull();
+  });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/claude-code-plugin/__tests__/config.test.ts]: Addressed - test renamed to 'does not detect circular references during config load'.

File: packages/cli/__tests__/integration/delegation-propagation.test.ts [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+  runCli,
+  getActiveState,
+  readRunbookState,
+} from '../helpers/test-utils.js';
+import { writeFile, readFile } from 'node:fs/promises';
+import { join } from 'node:path';
+
+describe('Delegation propagation integration', () => {
+  let workspace: TestWorkspace;
+
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/integration/delegation-propagation.test.ts:14]: Addressed - TestWorkspace type is now imported.

File: packages/cli/src/commands/abort.ts [hunk 4]
```diff
@@ -0,0 +1,275 @@
...
+          }
+
+          const { parentState, substepId } = scanResult;
+          const lock = new DelegationLock(cwd);
+
+          // 3. Acquire delegation lock
+          await lock.acquire(parentState.id);
+
+          let abortResult: ReturnType<typeof abortDelegation>;
+          let freshParent: RunbookState | null;
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/commands/abort.ts]: Addressed - freshParent is now initialized to null.

File: packages/cli/src/helpers/delegation-completion.ts [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+ * @returns 'handled' if propagation succeeded, 'stopped' if parent stopped,
+ *          'not-applicable' if no delegation linkage exists
+ */
+export async function handleDelegationCompletion(
+  childState: RunbookState,
+  result: 'pass' | 'fail',
+  cwd: string,
+  output: OutputEmitter,
+  depth = 0,
+): Promise<'handled' | 'stopped' | 'not-applicable'> {
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/helpers/delegation-completion.ts:61]: Addressed - @throws tag added to TSDoc.

File: packages/cli/src/helpers/delegation-completion.ts [hunk 6]
```diff
@@ -0,0 +1,199 @@
...
+  // Guard: no delegation linkage
+  if (!childState.delegation) {
+    return 'not-applicable';
+  }
+
+  // Guard: recursion limit
+  if (depth >= MAX_PROPAGATION_DEPTH) {
+    return 'handled';
+  }
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/helpers/delegation-completion.ts]: Addressed - console.warn now emitted when recursion limit reached.

File: packages/core/src/runbook/delegation-token.ts [hunk 7]
```diff
@@ -46,6 +46,30 @@ export function generateDelegationToken(): string {
   return TOKEN_PREFIX + encodeBase32(bytes);
 }
 
+/**
+ * Format a delegation token for safe display in output.
+ *
+ * Shows the prefix + first 3 base32 body chars + `...` + last 4 body chars.
+ * For example: `rdtk_ABC...4567`. This matches the design doc §11 security
+ * output hygiene: "first 7 + last 4 characters" where first 7 = `rdtk_` (5)
+ * + 3 body chars, last 4 = final 4 body chars.
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/delegation-token.ts]: Not applicable - current comment correctly says 'first 3 body chars'. The 'first 7' wording cited is not present.

File: packages/cli/__tests__/commands/abort.test.ts [hunk 8]
```diff
@@ -0,0 +1,355 @@
+import { describe, it, expect, beforeEach, afterEach } from '@jest/globals';
+import { createTestWorkspace, runCli, getActiveState, type TestWorkspace } from '../helpers/test-utils.js';
+import { writeFile } from 'node:fs/promises';
+import { join } from 'node:path';
+
+describe('abort command - unit tests', () => {
+  let workspace: TestWorkspace;
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/abort.test.ts]: Addressed in commit 7b04851.

File: packages/claude-code-plugin/__tests__/context.test.ts [hunk 9]
```diff
@@ -276,3 +276,244 @@ describe('Synthetic event context injection', () => {
...
+  test('follows priority order - subdirectory wins over flat when both exist', async () => {
+    const contextBase = path.join(testDir, '.claude', 'context');
+    const slashCommandDir = path.join(contextBase, 'slash-command');
+    await fs.mkdir(slashCommandDir, { recursive: true });
+
+    await fs.writeFile(path.join(slashCommandDir, 'test-command-start.md'), 'subdir');
+
+    const result = await discoverContextFile(testDir, 'test-command', 'start');
+    expect(result).toBe(path.join(slashCommandDir, 'test-command-start.md'));
+  });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/claude-code-plugin/__tests__/context.test.ts]: Addressed in 79ca097a - test now creates both flat and subdirectory files. Updated assertion to match actual priority (flat wins over subdirectory per buildContextPaths order).

File: packages/claude-code-plugin/__tests__/dispatcher.test.ts [hunk 10]
```diff
@@ -916,3 +916,307 @@ describe('gateMatchesFilePattern - debug logging', () => {
...
+  });
+
+  test('gateMatchesKeywords matches with extra whitespace', () => {
+    const gateConfig: GateConfig = {
+      command: 'npm test',
+      keywords: ['test'],
+    };
+
+    expect(gateMatchesKeywords(gateConfig, '  test  this  ')).toBe(true);
+  });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/claude-code-plugin/__tests__/dispatcher.test.ts]: Addressed in 79ca097a - consolidated into parameterized test.each, removed duplicate empty-string case already covered at line 295.

File: packages/cli/src/commands/abort.ts [hunk 11]
```diff
@@ -0,0 +1,275 @@
...
+
+            // 8. If force + childRunId: stop child run and record fail completion
+            if (options.force && childRunId) {
+              // Stop child run: capture linkage, delete state, pop session
+              const childState = await manager.load(childRunId);
+              if (childState) {
+                await manager.delete(childRunId);
+                const sessionService = new SessionService(manager);
+                await sessionService.popRunbook();
+              }
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/commands/abort.ts:262]: Addressed in 79ca097a - now extracts childState.agentId and passes to getActive(agentId) and popRunbook(agentId).

File: packages/cli/src/commands/claim.ts [hunk 12]
```diff
@@ -58,29 +59,20 @@ export function registerClaimCommand(program: Command): void {
...
-                if (delegationResult === 'stopped') shouldExitWithError = true;
+                const propResult = childState.variables.completed ? 'pass' : 'fail';
+                await handleDelegationCompletion(childState, propResult, cwd, output);
               }
             }
 
             output.flush();
-            if (shouldExitWithError) {
+            if (result.loopResult === 'stopped') {
               process.exit(1);
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/commands/claim.ts:85]: Addressed in 79ca097a - now captures handleDelegationCompletion return value and sets shouldExitWithError when propagation returns 'stopped'.

File: packages/cli/src/services/renderers/text-renderer.ts [hunk 13]
```diff
@@ -190,26 +190,38 @@ export class TextRenderer implements OutputRenderer {
...
+        frameKey?: string;
+        entry?: number;
+        unresolved?: number;
       };
+      step?: { name: string; description?: string };
+      lastAction?: { action: string; result?: boolean };
+      pending?: string[];
+      agents?: Record<string, { step: string; status: string; result?: string }>;
+      delegations?: { substep: string; runbook: string; state: string; childRunId?: string }[];
+    };
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/src/services/renderers/text-renderer.ts]: Addressed in 79ca097a - extracted to named StatusDetailData interface.

File: packages/core/src/output/zod-schemas.ts [hunk 14]
```diff
@@ -226,6 +226,28 @@ export const ActionResponseSchema = z
...
+    substep: z.string().describe('Substep ID owning the delegation'),
+    /** Path to the child runbook */
+    runbook: z.string().describe('Child runbook path'),
+    /** Current delegation state */
+    state: z
+      .enum(['pending', 'claimed', 'cancelled'])
+      .describe('Delegation state: pending, claimed, or cancelled'),
+    /** Child run ID (when claimed) */
+    childRunId: z.string().optional().describe('Child run ID when delegation is claimed'),
+  })
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/output/zod-schemas.ts:248]: Addressed in 79ca097a - added .refine() requiring childRunId when state is 'claimed'.

File: packages/core/src/runbook/delegation-service.ts [hunk 15]
```diff
@@ -11,6 +11,26 @@ import type {
...
+  readonly force?: boolean;
+}
+
+/**
+ * Result of aborting a delegation.
+ */
+export type AbortDelegationResult =
+  | { readonly status: 'cancelled'; readonly updatedSubstepStates: readonly SubstepState[] }
+  | { readonly status: 'already_cancelled' }
+  | { readonly status: 'needs_force'; readonly childRunId: string };
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/src/runbook/delegation-service.ts]: Addressed in 79ca097a - decomposed into documented interfaces (AbortDelegationCancelledResult, AbortDelegationAlreadyCancelledResult, AbortDelegationNeedsForceResult) with field-level JSDoc.

File: packages/cli/__tests__/commands/abort.test.ts [hunk 16]
```diff
@@ -0,0 +1,361 @@
...
+
+      // First abort should succeed
+      const result1 = runCli(`abort ${token}`, workspace);
+      expect(result1.exitCode).toBe(0);
+
+      // Second abort should be idempotent
+      const result2 = runCli(`abort ${token}`, workspace);
+      expect(result2.exitCode).toBe(0);
+      expect(result2.stdout).toMatch(/already cancelled/i);
+    });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/abort.test.ts]: Acknowledged - trivial nitpick, will address in a follow-up if needed.

File: packages/cli/__tests__/commands/abort.test.ts [hunk 17]
```diff
@@ -0,0 +1,361 @@
...
+
+    it('abort with multiple tokens uses only first', async () => {
+      const token = await setupDelegation();
+
+      // Pass multiple tokens - Commander should only use first positional arg
+      const result = runCli(`abort ${token} extra-arg`, workspace);
+      // Commander may error on unexpected argument
+      // This documents the behavior
+      expect(result.exitCode).not.toBe(undefined);
+    });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/abort.test.ts:273]: Acknowledged - trivial nitpick, will address in a follow-up if needed.

File: packages/cli/__tests__/commands/abort.test.ts [hunk 18]
```diff
@@ -0,0 +1,361 @@
...
+    it('handles abort during runbook execution', async () => {
+      const token = await setupDelegation();
+
+      // Start claim but don't complete it
+      runCli(`claim ${token}`, workspace);
+
+      // Force abort while execution is in progress
+      const result = runCli(`abort ${token} --force`, workspace);
+      expect(result.exitCode).toBe(0);
+    });
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/abort.test.ts:346]: Acknowledged - trivial nitpick, will address in a follow-up if needed.

File: packages/core/__tests__/runbook/abort-delegation.test.ts [hunk 19]
```diff
@@ -0,0 +1,110 @@
...
+    retryCount: 0,
+    variables: {},
+    steps: [{ id: '1', status: 'running' }],
+    pendingSteps: [],
+    agentBindings: {},
+    startedAt: '2026-02-27T10:00:00.000Z',
+    updatedAt: '2026-02-27T10:00:00.000Z',
+    substepStates,
+  } as RunbookState;
+}
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/__tests__/runbook/abort-delegation.test.ts:35]: Acknowledged - trivial nitpick, will address in a follow-up if needed.

File: packages/core/__tests__/runbook/format-token-hint.test.ts [hunk 20]
```diff
@@ -0,0 +1,36 @@
+import { describe, it, expect } from '@jest/globals';
+import { truncateDelegationToken, TOKEN_PREFIX } from '../../src/runbook/delegation-token.js';
+
+describe('truncateDelegationToken', () => {
+  it('truncates a standard 37-char token to prefix + first 3 + ... + last 4', () => {
+    // rdtk_ + 32 base32 chars
+    const token = `${TOKEN_PREFIX}ABCDEFGHIJKLMNOPQRSTUVWXYZ234567`;
+    const hint = truncateDelegationToken(token);
+    expect(hint).toBe('rdtk_ABC...4567');
```
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/__tests__/runbook/format-token-hint.test.ts]: Acknowledged - trivial nitpick, will address in a follow-up if needed.

File: packages/cli/__tests__/commands/abort.test.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/abort.test.ts]: Addressed in 9861dc8d.

File: packages/cli/__tests__/commands/abort.test.ts (hunk 17 above)
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/abort.test.ts:273]: Addressed in 9861dc8d.

File: packages/cli/__tests__/commands/abort.test.ts (hunk 18 above)
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/cli/__tests__/commands/abort.test.ts:346]: Addressed in 9861dc8d.

File: packages/core/__tests__/runbook/abort-delegation.test.ts (hunk 19 above)
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/__tests__/runbook/abort-delegation.test.ts:35]: Addressed in 9861dc8d.

File: packages/core/__tests__/runbook/format-token-hint.test.ts (hunk 20 above)
[the remark this answers is not in the document]
Comment (tobyhede) [reply packages/core/__tests__/runbook/format-token-hint.test.ts]: Addressed in 9861dc8d.
